src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala

```diff
@@ -1669,6 +1669,7 @@ abstract class SpecializeTypes extends InfoTransform with TypingTransformers {
               val tree1 = addBody(ddef, target)
               (new ChangeOwnerTraverser(target, tree1.symbol))(tree1.rhs)
               debuglog("changed owners, now: " + tree1)
+              if (target.hasFlag(Flags.SYNCHRONIZED)) ddef.symbol.setFlag(Flags.SYNCHRONIZED)
```

Comment:
Good question. I've just poked around and found that the `Function0.apply` is used as the prototype from with `Foo.apply$mc...` is cloned. I've found a better place to do this and force pushed a new version.